contribs/gnodev/pkg/dev/node_state_test.go

```diff
@@ -0,0 +1,189 @@
+	"github.com/gnolang/gno/contribs/gnodev/pkg/events"
+	"github.com/gnolang/gno/gno.land/pkg/gnoclient"
+	"github.com/gnolang/gno/gno.land/pkg/gnoland"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestNodeMovePreviousTX(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	const callInc = 5
+
+	node, emitter := testingCounterRealm(t, callInc)
+
+	t.Run("Prev TX", func(t *testing.T) {
+		err := node.MoveToPreviousTX(ctx)
+		require.NoError(t, err)
+		assert.Equal(t, events.EvtReload, emitter.NextEvent().Type())
+
+		// Check for correct render update
+		render, err := testingRenderRealm(t, node, "gno.land/r/dev/counter")
+		require.NoError(t, err)
+		require.Equal(t, render, "4")
+	})
+
+	t.Run("Next TX", func(t *testing.T) {
+		err := node.MoveToNextTX(ctx)
+		require.NoError(t, err)
+		assert.Equal(t, events.EvtReload, emitter.NextEvent().Type())
+
+		// Check for correct render update
+		render, err := testingRenderRealm(t, node, "gno.land/r/dev/counter")
+		require.NoError(t, err)
+		require.Equal(t, render, "5")
+	})
+
+	t.Run("Multi Move TX", func(t *testing.T) {
+		moves := []struct {
+			Move           int
+			ExpectedResult string
+		}{
+			{-2, "3"},
+			{2, "5"},
+			{-5, "0"},
+			{5, "5"},
+			{-100, "0"},
+			{100, "5"},
+			{0, "5"},
+		}
+
+		t.Logf("initial state %d", callInc)
+		for _, tc := range moves {
+			t.Logf("moving from `%d`", tc.Move)
+			err := node.MoveFrom(ctx, tc.Move)
+			require.NoError(t, err)
+			if tc.Move != 0 {
+				assert.Equal(t, events.EvtReload, emitter.NextEvent().Type())
+			}
+
+			// Check for correct render update
+			render, err := testingRenderRealm(t, node, "gno.land/r/dev/counter")
+			require.NoError(t, err)
+			require.Equal(t, render, tc.ExpectedResult)
+		}
+	})
+}
+
+func TestSaveCurrentState(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	node, emitter := testingCounterRealm(t, 2)
+
+	// Save current state
+	err := node.SaveCurrentState(ctx)
+	require.NoError(t, err)
+
+	// Send a new tx
+	msg := gnoclient.MsgCall{
+		PkgPath:  "gno.land/r/dev/counter",
+		FuncName: "Inc",
+		Args:     []string{"10"},
+	}
+
+	res, err := testingCallRealm(t, node, msg)
+	require.NoError(t, err)
+	require.NoError(t, res.CheckTx.Error)
+	require.NoError(t, res.DeliverTx.Error)
+	assert.Equal(t, events.EvtTxResult, emitter.NextEvent().Type())
+
+	// Test render
+	render, err := testingRenderRealm(t, node, "gno.land/r/dev/counter")
+	require.NoError(t, err)
+	require.Equal(t, render, "12") // 2 + 10
+
+	// Reset state
+	err = node.Reset(ctx)
+	require.NoError(t, err)
+	assert.Equal(t, events.EvtReset, emitter.NextEvent().Type())
+
+	render, err = testingRenderRealm(t, node, "gno.land/r/dev/counter")
+	require.NoError(t, err)
+	require.Equal(t, render, "2") // Back to the original state
+}
+
+func TestExportState(t *testing.T) {
```

Comment:
Parallel tests are cooler than sequential tests :)

Reply:
Same as the other test, I'm still unsure whether to run node tests in parallel and prefer to avoid it to prevent inconsistency and false-positive tests.